src/System.Runtime.Extensions/src/System/AppDomain.cs

```diff
@@ -280,5 +286,45 @@ public void SetThreadPrincipal(IPrincipal principal)
                 _defaultPrincipal = principal;
             }
         }
+
+        public IPrincipal GetThreadPrincipal()
+        {
+            IPrincipal principal = _defaultPrincipal;
+            if (principal == null)
+            {
+                switch (_principalPolicy)
+                {
+                    case PrincipalPolicy.UnauthenticatedPrincipal:
+                        if (s_genericPrincipalCtor == null || s_genericIdentityCtor == null)
+                        {
+                            Assembly assembly = Assembly.Load(new AssemblyName("System.Security.Claims"));
+                            Type genericPrincipal = assembly.GetType("System.Security.Principal.GenericPrincipal", throwOnError: true);
+                            Type genericIdentity = assembly.GetType("System.Security.Principal.GenericIdentity", throwOnError: true);
+
+                            s_genericPrincipalCtor = genericPrincipal.GetConstructor(new[] { genericIdentity, typeof(string[]) });
+                            s_genericIdentityCtor = genericIdentity.GetConstructor(new[] { typeof(string), typeof(string) });
+                        }
+
+                        principal = (IPrincipal)s_genericPrincipalCtor.Invoke(new object[] { s_genericIdentityCtor.Invoke( new object[] { string.Empty, string.Empty }), new string[] { string.Empty } });
```

Comment:
There's no need to cache the arrays used to initialize the s_ members, but for these arrays that are used on every call and that aren't based on inputs, e.g.
```C#
new object[] { string.Empty, string.Empty }
new string[] { string.Empty }
```
we should consider caching those as well.

Reply:
This is called once per thread. I do not think caching the arrays is worth it.

If we wanted to make this leaner compared to what it is now, we should cache delegate to method to create the principal (it would fix the problem with the arrays as side-effect too).

Reply:
> If we wanted to make this leaner compared to what it is now, we should cache delegate to method to create the principal (it would fix the problem with the arrays as side-effect too).

Let's do that :)

Reply:
Here is what I had in mind:

```csharp
Type type = Type.GetType("System.Security.Principal.GenericPrincipal, System.Security.Claims", throwOnError: true);
s_getUnauthenticatedPrincipal = (Func<IPrincipal>)mi.CreateDelegate(typeof(Func<IPrincipal>), type, "GetDefaultInstance");
```

And add `GetDefaultInstance` to `System.Security.Principal.GenericPrincipal`.